Add renamer phase and a simala backend for natural4

fendor opened the pull request:
This PR introduces two new things:

* A Renamer/Name resolver phase for `Rule`s
* A Transpiler which allows transpiling a `Rule` to a Simala `Decl`.

The renamer implements lexical scope checking for `Hornlike` and `TypeDecl` rules.
It resolves all names, replacing occurrences of each name with its renamed equivalent.
This is supposed to lay the groundwork for a principled transpilation pipeline, avoiding redoing work for each transpiler, such as resolving names, making sure variables are in scope, etc...
In the future, we want to implement a type checker on top of a `RnRule`. However, that's currently only a dream and there are no attempts of doing any kind of typechecking at the moment.

The simala transpiler solely operates on the renamed rules. This simplifies the implementation a lot, demonstrating how other transpilers may also benefit from the rename phase.
The simala backend can currently only handle `Hornlike` rules and ignores any kind of `TypeDecl` rules, as simala is dynamically typed and typechecking should occur in a dedicated rule phase, not ad-hoc in the transpiler.

Both components are somewhat experimental and are expected to have bugs and missing features.

Todos:
* [x] Logging (potentially using colog style logging)
* [x] A proper error type (Don't throw `String`s)
* [x] Integration with the main-cli

fendor commented on lib/haskell/hie.yaml:
```diff
@@ -1,37 +1,3 @@
-      component: "natural4:bench:natural4-bench"
-
-    - path: "natural4/bnfc"
-      component: "natural4:exe:l4-bnfc-exe"
+  cabal:
+    component: all
```
TODO: undo this, this is only for my local development :see_no_evil:

inariksit commented on lib/haskell/natural4/src/LS/Renamer.hs:
```diff
@@ -0,0 +1,1295 @@
+  | FixArityFunctionNotFound RnName [RnExpr]
+  | ArityErrorLeft !Int RnName [RnExpr]
+  | ArityErrorRight !Int RnName [RnExpr]
+  | UnexpectedNameNotFound OccName
+  | UnexpectedRnNameNotFound RnName
```
Minor thing, but as I read this I was wondering what are OccName and RnName, and what's the difference between these two errors. I'm sure I'll find the answers elsewhere, but just reporting that a comment here could be useful.

ym-han commented:
Some quick feedback:

## Renamer.hs

High level things
* Would be helpful to have a comment or two indicating how to use the Renamer (might be enough to just indicate what the main entrypoints are in the export list and link to the Simala transpiler as an example?)
* Might be helpful to have a quick comment that 
(i) summarizes what the assumptions and guarantees of the Renamer are
(ii) summarizes why the renaming is sound, if there is an easy way to explain this at a high level --- there may not

Specific things
* `RnRelationalPredicate`: Might be good to provide examples of the various variants, especially `RnRelationalTerm`. This isn't high-priority / doesn't have to be done in this PR. 
* `* @x@: a variable, might be bound ad-hoc` -- what does 'bound ad-hoc' mean in this context?

inariksit commented on lib/haskell/natural4/test/testdata/golden/xpile/simala/rodents-and-vermin.expected:
```diff
@@ -0,0 +1,2 @@
+Failed transpilation:
+Unsupported RelationalPredicate: RPis
```
Is this the excpected result?

fendor replied:
Yes! This is intended as documenting the current capabilties of the transpilation.
I struggled to identify all possible meanings of predicates. In this case, 
```
"Loss or Damage" IS ANY ( "caused by rodents"
                                    , "caused by insects"
                                    , "caused by vermin"
                                    , "caused by birds"
                                    )
```

Is a weird RelationalPredicate in my opinion. `ANY` feels overloaded. It may mean `OR`, but `IS ANY` suddenly means `ELEM`, and there is actually a `RPelem` `RelationalPredicate` which we don't use here.

So, I didn't even try to support this example, but rather we should talk about, which composite operators we want to support, or whether we rather want to use some `ELEM` operator.

inariksit replied:
Good point! I pretty much have forgotten that RPelem existed 😁

ym-han replied:
Yeah in general I think we shouldn't be afraid to try to make the syntax more regimented and understandable (not least because there is a non-trivial chance we'll actually have to try to help other people learn and use this!)

ym-han commented on lib/haskell/natural4/src/LS/XPile/MathLang/GenericMathLang/TranslateL4.hs:
```diff
@@ -474,7 +480,7 @@ mkL4VarTypeDeclAssocList = convertL4Types . declaredVarsToAssocList
 
     convertL4Types :: [(T.Text, Maybe TypeSig)] -> [(Var, Maybe L4EntityType)]
-    convertL4Types = each % _1 %~ mkVar >>> each % _2 %~ fmap mkEntType
+    convertL4Types = fmap (bimap mkVar (fmap mkEntityType))
```
Actually I'm curious now: are the differences mere stylistic (lol), or are there, e.g., performance reasons to prefer one of the versions?

fendor replied:
Haha, this is purely stylistic, and I only accidentally committed this line, which I personally think is more readable. But this is a remnant from reading the `TranslateL4` code, making notes for myself, not intended as an improvement of anything.
While there may be differences in performance, I don't think it is relevant to the performance of the program. Also, I did not measure or look at the core, so all versions may compile to the same final result, who knows :)

ym-han commented on lib/haskell/natural4/src/LS/XPile/Simala/Transpile.hs:
```diff
@@ -0,0 +1,881 @@
+
+litToSimala :: RnLit -> Simala.Expr
+litToSimala = \case
+  RnInt int -> Simala.Lit $ Simala.IntLit $ fromIntegral int -- TODO: why does simala only support 'Int'?
```
I think Andres just hadn't gotten around to adding more numeric types

fendor replied:
I talked to @kosmikus about this and the reason is :shrug: (yes, a literal shoulder shrug)

It didn't matter at the time, and the implementation should be driven by requirements. If we really want and need infinite sized integers, it is still trivial to change the `Simala.IntLit` type to accept `Integer`.

fendor commented on lib/haskell/natural4/src/LS/Renamer.hs:
```diff
@@ -0,0 +1,1295 @@
+renameRule _ r@Rule.RegBreach{} = throwError $ UnsupportedRule "renameRule" r
+renameRule _ r@Rule.NotARule{} = throwError $ UnsupportedRule "renameRule" r
+
+renameLocalRules :: Tracer Log -> [Rule] -> Renamer [RnRule]
+renameLocalRules = renameRules
```
So, in lexical scoping, we have a hierarchy of scopes. `GIVETH` and `DECIDE` are not "global", but their definitions may be used one level up the hierarchy. This generalises nicely, as rules in `WHERE` can be used by `DECIDE` and other rules in the `WHERE` clause, but not outside the rule that defines the `WHERE` clause.

For a practical example with nested clauses:

```
GIVEN x
DECIDE f x IS y
WHERE
  y IS g x
  GIVEN d DECIDE g d IS y WHERE y IS SUM(d, d)
```

* Then the rules `y IS g x` and `GIVEN d DECIDE g d IS y WHERE y IS SUM(d, d)` is local to the rule and the function `g` may be used in both `WHERE` and `DECIDE` blocks.
* `y IS SUM(d,d)` is local to the rule `g`, and can not be referenced by the top-level `WHERE` or `DECIDE`.
* `y IS SUM(d,d)` name shadows the outer `y`.

inariksit replied:
I see, thanks for the explanation!

ym-han replied:
This is another tangential syntax thing, but it's not clear to me that we really should allow users the flexibility to write arbitrary declarations in a `WHERE` block (as opposed to just things of the form `y IS g x`. There's probably no easy way to answer this question right now, and it's definitely not high-priority, but I'd be curious to see what you all think.

fendor commented on lib/haskell/natural4/src/LS/Renamer.hs:
```diff
@@ -0,0 +1,1295 @@
+          | isL4BuiltIn name -> do
+              -- ANDRES: I'm not convinced that built-ins should be renamed, and
+              -- if we already detected that they're built-ins, perhaps we should
+              -- just use a different dedicated constructor for this case.
```
Fixed!

fendor commented:
I tried to address some of the feedback, moved some code around that makes it hopefully a bit more clear what we are trying to do and added a documentation block that will hopefully evolve to a specification of the scoping rules in naturalL4.

A lot of good questions were raised regarding supported syntax and the type system, which we should write down and discuss.

I also fixed some glaring issues with the Simala Transpiler. However, I expect there are many issues left in the code, which we will hopefully weed out over time.
The renamer seems to be working relatively well, but is likely not tested enough. It would be nice to come up with a test specification that is better than simply printing a renamed rule, as this is really difficult to check, verify and understand.

ym-han commented on lib/haskell/natural4/src/LS/Renamer.hs:
```diff
@@ -0,0 +1,1295 @@
+  LS.RPBoolStructR lhs _predicate _rhs -> do
+    scanDecideMultiTerm tracer lhs
+  LS.RPnary LS.RPis (lhs : _rhs) -> do
+    -- When the assignment has multiple complicated relational predicates,
```
this kind of comment is quite helpful, thanks!

ym-han commented on lib/haskell/natural4/src/LS/Renamer.hs:
```diff
@@ -0,0 +1,1295 @@
+-- Even in that case, since 'TypeDecl''s 'has' is a list of 'TypeDecl''s, it seems like
+-- there is no arbitrary nesting.
+--
+-- ANDRES: I think the fact that type signatures allow nested
```
Agreed

ym-han commented on lib/haskell/natural4/src/LS/Renamer.hs:
```diff
@@ -0,0 +1,1295 @@
+
+-- | Rename a 'LS.MultiTerm' and turn each 'LS.MTExpr' into a 'RnExpr'.
+--
+-- Renaming a list of 'LS.MTExpr' cannot be done without keeping intermediate
```
Again, very helpful comments!

ym-han reviewed (approved):
Thanks for adding more docs / explanations!

fendor commented on lib/haskell/natural4/src/LS/Renamer.hs:
```diff
@@ -0,0 +1,1295 @@
+{-# LANGUAGE DerivingVia #-}
+{-# LANGUAGE DuplicateRecordFields #-}
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE LambdaCase #-}
```
TBH, I just added extensions as needed :see_no_evil: But I am cleaning them up now, as I don't need all of them!

ym-han replied:
No worries, I do that too. Was mostly asking out of curiosity / because I wanted to learn good Haskell practices
